presto-main/src/main/java/com/facebook/presto/execution/scheduler/TopologyAwareNodeSelector.java

```diff
@@ -173,17 +184,48 @@ public Node selectCurrentNode()
                 assignmentStats.addAssignedSplit(chosenNode);
                 topologicCounters[chosenDepth]++;
             }
+            else {
+                splitWaitingForAnyNode = true;
+            }
         }
         for (int i = 0; i < topologicCounters.length; i++) {
             if (topologicCounters[i] > 0) {
                 topologicalSplitCounters.get(i).update(topologicCounters[i]);
             }
         }
-        return assignment;
+
+        CompletableFuture<?> blocked;
+        if (existingTasks.isEmpty()) {
+            blocked = completedFuture(null);
+        }
+        else {
+            int maxPendingForWildcardNetworkAffinity = calculateMaxPendingSplits(0, networkLocationSegmentNames.size());
+            if (splitWaitingForAnyNode) {
+                List<CompletableFuture<?>> stateChangeFutures = existingTasks.stream()
+                        .map(remoteTask -> remoteTask.whenSplitQueueHasSpace(calculateLowWatermark(maxPendingForWildcardNetworkAffinity)))
+                        .collect(toImmutableList());
+                blocked = firstCompletedFuture(stateChangeFutures, true);
+            }
+            else {
+                blocked = toWhenHasSplitQueueSpaceFuture(blockedExactNodes, existingTasks, calculateLowWatermark(maxPendingForWildcardNetworkAffinity));
+            }
+        }
+        return new SplitPlacementResult(blocked, assignment);
+    }
+
+    private int calculateMaxPendingSplits(int currentDepth, int totalDepth)
```

Comment:
Can you explain what this function is trying to compute? Specifically, what's the underlying model it represents? 

For, say, totalDepth = 3, is f(0, 3) + f(1, 3) + f(2, 3) + f(3, 3) supposed to add up to maxPendingSplitsPerTask? From the code, it appears not to be the case, hence my confusion about what it's meant to do.

The comment below `double queueFraction = ...` doesn't help clarify things, unfortunately.